kong/plugins/oauth2/access.lua

```diff
@@ -88,6 +88,10 @@ do
   end
 end
 
+local function get_plugin_identifier(conf)
+  return type(conf) == "table" and conf.__key__ or nil
```

Comment:
Is it normal for `conf` to be nil or for `conf.__key__` to be nil, or should that case be handled as an error?

Reply:
I believe that `conf` will always be a table and that `conf.__key__` will always be set.

However I have doubts about whather `conf.__key__` is appropiate for this.

By looking at how it's set, it seems [the plugins iterator](https://github.com/Kong/kong/blob/master/kong/runloop/plugins_iterator.lua#L161) sets it the first time it loads a configuration of a plugin.

Perhaps I am reading that code incorrectly. What I am reading is:

* The order in which requests are processed conditions `conf.__key__` (if I have two routes A and B with two plugins PA and PB, if I make a request to A first and then to B, `PA.__key__` will be `1` and `PB.__key__` will be `2`. If I do the requests in the opposite order, the keys will also be reversed.
* Because of this, different workers may assign different `conf.__key__` values to the same plugin instances (one worker can treat A first, while one worker can treat B first).
* For the same reason, the same worker could assign different `conf.__key__` values after a `kong reload`.

In short, `__key__` is only appropriate for 1-worker, non-persisted data. The plugin identifier, however, is persisted on the `oauth2_authorization_codes` table.

I think we either need to make `__key__` significantly more complicated (possible candidate: we make it a hash of  `inspect(conf)` -- inspect sorts keys alphabetically and recursively) or we need to use something other than `__key__` here.

Reply:
Hi @kikito , based on my understanding, I think the `conf.__key__` value will not be related to the request order or worker id. The value was only combined by several entity ids, see [here](https://github.com/Kong/kong/blob/master/kong/runloop/plugins_iterator.lua#L154). Since Kong won't allow a plugin to be executed multiple times. So the `conf.__key__` could represent the "identifier" of the plugin.

Another implementation is to use the `plugin.id` as plugin's identifier. Unfortunately, it's currently unable to retrieve its id(uuid) in the plugin handler. I do mention this on another PR(https://github.com/Kong/kong/pull/9903#discussion_r1065331867)

Reply:
You are right, I mistook `__seq__` with `__key__`. It was late last night, apologies.

I then agree with @flrgh . The plugin should fail greatly if for some reason `conf` is not a table (or `conf.__key__`). I suggest we change the line to

```suggestion
  return conf.__key__
```

Or just use `conf.__key__` below

Reply:
SGTM